// resolve-dependencies/lib/node-loader.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.load = exports.resolve = exports.resolveSync = void 0;
const fs = __importStar(require("fs"));
const globby_1 = require("globby");
const path_1 = require("path");
const gather_deps_1 = require("./gather-deps");
const file_1 = require("./file");
const enhanced_resolve_1 = require("enhanced-resolve");
const { readFileSync, lstatSync, statSync, realpathSync } = fs;
const fileSystem = new enhanced_resolve_1.CachedInputFileSystem(fs, 4000), resolver = enhanced_resolve_1.ResolverFactory.createResolver({
    extensions: ['.js', '.json', '.node'],
    symlinks: false,
    fileSystem: new enhanced_resolve_1.CachedInputFileSystem(fs, 4000),
}), syncResolver = enhanced_resolve_1.ResolverFactory.createResolver({
    extensions: ['.js', '.json', '.node'],
    useSyncFileSystemCalls: true,
    symlinks: false,
    fileSystem,
}), defaultOptions = { loadContent: true, expand: 'none', isEntry: false };
function resolveSync(from, request) {
    const result = {
        absPath: '',
        pkgPath: '',
        pkg: null,
        warning: '',
    };
    syncResolver.resolve({}, from, request, {}, (err, path, data) => {
        if (err) {
            result.warning = err.message;
            return;
        }
        result.absPath = path;
        result.pkgPath = data.descriptionFilePath;
        result.pkg = data.descriptionFileData;
        return;
    });
    return result;
}
exports.resolveSync = resolveSync;
function resolve(from, request) {
    const result = {
        absPath: '',
        pkgPath: '',
        pkg: null,
        warning: '',
    };
    return new Promise((resolve) => {
        resolver.resolve({}, from, request, {}, (err, path, data) => {
            if (err) {
                result.warning = err.message;
                return resolve(result);
            }
            result.absPath = path;
            result.pkgPath = data.descriptionFilePath;
            result.pkg = data.descriptionFileData;
            resolve(result);
        });
    });
}
exports.resolve = resolve;
async function expand(file, fileDir, baseDir, globs) {
    const files = globby_1.sync(globs, {
        cwd: baseDir,
        followSymbolicLinks: false,
    });
    files
        .sort() //glob search is width first and not deterministic
        .map((dep) => file_1.ensureDottedRelative(fileDir, path_1.join(baseDir, dep)))
        .filter((relDep) => file.absPath !== path_1.join(baseDir, relDep))
        .forEach((relDep) => {
        file.deps[relDep] = file.deps[relDep] || null;
    });
    const currentDeps = Object.keys(file.deps);
    file.package &&
        file.package.dependencies &&
        Object.keys(file.package.dependencies || {}).forEach((dependency) => {
            if (!currentDeps.some((curDep) => curDep.startsWith(dependency))) {
                file.deps[dependency] = file.deps[dependency] || null;
            }
        });
}
function load(workingDirectory, request, options = defaultOptions) {
    var _a;
    const { absPath, pkg, pkgPath, warning } = resolveSync(workingDirectory, request);
    if (!absPath) {
        return { warning: warning };
    }
    const file = file_1.createFile(absPath), isJs = absPath.endsWith('.js') || absPath.endsWith('.mjs') || options.isEntry;
    file.absPath = absPath;
    if (isJs || absPath.endsWith('json')) {
        file.contents = readFileSync(absPath, 'utf-8');
    }
    if (isJs) {
        try {
            const parseResult = gather_deps_1.gatherDependencies(file.contents, absPath.endsWith('.mjs'));
            Object.assign(file.deps, parseResult.deps);
            file.variableImports = parseResult.variable;
        }
        catch (e) {
            return { warning: `Error parsing file: "${file.absPath}"\n${e.stack}` };
        }
    }
    const fileDir = path_1.dirname(file.absPath), expandVariable = Boolean(options.expand === 'variable' && file.variableImports);
    if (file_1.isNodeModule(request) && pkg && pkgPath) {
        file.package = pkg;
        file.deps[file_1.ensureDottedRelative(fileDir, pkgPath)] = null;
        const pkgDir = (file.moduleRoot = path_1.dirname(pkgPath)), expandAll = options.expand === 'all';
        if (expandVariable || expandAll) {
            expand(file, fileDir, pkgDir, file_1.nodeModuleGlobs(file));
            file.contextExpanded = true;
        }
        if (file_1.extraGlobs(file).length) {
            expand(file, fileDir, pkgDir, file_1.extraGlobs(file));
        }
    }
    else if (expandVariable && ((_a = options.context) === null || _a === void 0 ? void 0 : _a.moduleRoot) && !options.context.expanded) {
        expand(file, fileDir, options.context.moduleRoot, file_1.nodeModuleGlobs({ package: { files: options.context.globs } }));
        file.contextExpanded = true;
    }
    if (!options.loadContent) {
        file.contents = null;
    }
    const stats = lstatSync(file.absPath);
    if (stats.isSymbolicLink()) {
        const path = realpathSync(file.absPath);
        const absStat = statSync(file.absPath);
        file.realPath = path;
        file.realSize = absStat.size;
    }
    file.size = stats.size;
    return file;
}
exports.load = load;
